src/analysis/souffle/taint.dl

```diff
@@ -70,21 +83,30 @@
 // that source path. Instead, we remove all of the tags on the source from the
 // midpoint access path we add in the middle of each edge. This rule populates a
 // request to remove a tag on the edge midpoint.
-//      - it is safe because these requests are only accepted if they would 
+//      - it is safe because these requests are only accepted if they would
 //      normally be by the removeTag(_, _) rule
-//      - it does not require the manifest writer to know anything about the 
+//      - it does not require the manifest writer to know anything about the
 //      tags in the analysis
-//      - data owners do not need to know about paths, and they will grant 
-//      permission for edge removal indirectly by granting permission to remove 
+//      - data owners do not need to know about paths, and they will grant
+//      permission for edge removal indirectly by granting permission to remove
 //      tags (abstract representations of their data).
 
-saysRemoveTag(principal, midpoint, tag) :-
-  isTag(tag), claimNotEdge(principal, src, tgt), edgeToMidpointAccessPath(src, tgt, midpoint).
+ownsAccessPath(principal, path) :- saysOwnsAccessPath(principal, principal, path).
+ownsAccessPath(principal, tgt) :-
+  resolvedEdge(principal, src, tgt), ownsAccessPath(principal, src).
```

Comment:
See the changes in [tag_claim.h](https://github.com/google-research/raksha/pull/215/files#diff-d63a9c49012e0f82e22642cd080442fc151b7916ca2fc1a5d8fa5b9621a33499), where we use this rule for a claim. A `claim x is "foo"` in particle "P" would be translated as follows:
```
says_hasTag("P", "x", owner, "foo") :- ownsAccessPath(owner, "x").
```

Conceptually, `ownsAccessPath` is just another tag on the access path captured in a separate relation.

Reply:
> So I don't see how this rule stops anything from happening. 

Meaning that there is no way for `ownsAccessPath(principal, path)` is true? 

`ownsAccessPath` is just another tag that is captured as a relation. You are right that there is some missing delegation w.r.t. ownsAccessPath. The way I envision it, there would be some sort of root authority which validates `saysOwnsAccessPath(principal, path)` statements. The root authority will be like the trusted root certificate in a system.

Reply:
> Meaning` that there is no way for ownsAccessPath(principal, path) is true?

No I was saying the opposite: because of the rule on line 94 which says: `ownsAccessPath(principal, path) :- saysOwnsAccessPath(principal, principal, path).` it's always possible for a principal to take ownership of a path. If the goal of a feature is to improve security, it should either: 1) make "insecure" executions impossible, or 2) make mistakes leading to insecurity harder. I was saying I don't think it does (1), at least.

What I was saying is not entirely true though, because:

```
says_hasTag("P", "x", owner, "foo") :- ownsAccessPath(owner, "x").
```

means that for `saysHasTag(P, x, owner, foo)` can only be made true if the owner of `"foo"` decides to take ownership of the path `"x"`, so an arbitrary principal "`P`" can't make this rule fire.

I still don't think we need this though because in our implementation [as of this commit](https://github.com/google-research/raksha/blob/5fb285008ef8232a3e82fc0010b61609606e65fc/src/analysis/souffle/taint.dl)
we had a rule that said `hasTag(tgt, tag) :- claimHasTag(principal, tgt, tag), ownsTag(principal, tag)` (and we recently replaced `claimHasTag(principal, _, _)` with just `saysHasTag(principal, _, _)`). If we go back to this rule, it already prevents tags from being added to paths without permission from the owner of the tag. (And I think even in our current snapshot we already have a rule that's similar to this).

Reply:
Just summarizing the discussion here. When writing a particle, we will not know all the users that will be feeding data to that particle. Therefore, we cannot have the following claims in a particle. 
```
claim foo is "userA:userSelection"
claim foo is "userB:userSelection"
```

The way we capture the users whose data propagates through that access path is through `ownsAccessPath`. It is not different from a tag except that we capture it using a dedicated relation. In fact, if we can have a special tag called `"ownership"`, then we can have the following inference rule:
```
ownsAccessPath(foo, owner) :- hasTag(foo, owner, "ownership").
```
Given this discussion, we use `ownsAccessPath` to handle `claim foo is "userSelection"` in particle `"P1"` as follows:

says_hasTag("P1",  "foo", owner, "userSelection") :- ownsAccessPath("foo", owner).

Reply:
Sorry to be a pain. I still think we can cut out the `ownsAccessPath` relation because it is redundant and I think it does not solve the problem of not knowing the users beforehand. After this comment, I will stop being a blocker, though.

Why I think it's redundant:
I think your rule for `mayHaveTag` on line 108 of `taint.dl` already propagates does the propagation of ownership information that the rules for ownsAccessPath on lines 94-96 are also doing. During the meeting, I thought the rules of the form `says_hasTag(...) :- ownsAccessPath(...)` may have been needed to check that owners agree to adding a tag, but the rule for `hasTag` on line 104 of `taint.dl` does this because the speaker and owner have to be the same on the RHS of the rule.

Why I think it doesn't help with not knowing the users a priori:
To make the rules of the form `says_hasTag(...) :- ownsAccessPath(...)` fire, you still need to also generate predicates of the form `ownsAccessPath(path, owner)` where the specific owner (UserA? UserB?) is included. I think if you can generate this `ownsAccessPath(path, owner)` predicate, you could also just generate `saysHasTag(speaker, path, owner, tag)` directly.

Reply:
Note that `ownsAccessPath` does not introduce a new concept. It is just a new tag (say `"ownership"`) that is captured using the `ownsAccessPath` relation for _convenience_. We can capture an `ownsAccessPath(owner, tag)` with `mayHaveTag` if we have a designated tag like `"ownership"` as follows:

```
ownsAccessPath(owner, foo) :- mayHaveTag(foo, owner, "ownership").
```

> To make the rules of the form says_hasTag(...) :- ownsAccessPath(...) fire, you still need to also generate predicates of the form ownsAccessPath(path, owner) where the specific owner (UserA? UserB?) is included. 

Yes, that is right. However, we will  only claim `ownsAccessPath` on access path at the point of data ingestion. The `ownsAccessPath` propagation rule will propagate `ownsAccessPath` along the data flow edges.

(Just an aside: we will need to do the same sort of delegation as `hasTag` for `ownsAccessPath`, but this draft PR does not take care of it. However, that shouldn't invalidate any of my comments here.)

> I think if you can generate this ownsAccessPath(path, owner) predicate, you could also just generate saysHasTag(speaker, path, owner, tag) directly.

No that will not be sufficient. Let us say we have a particle "P" that has a `claim "foo" is "userSelection"` and the "EndUser" directly delegates to particle "P":
```
// "EndUser" trusts "P" to add tags correctly. 
"EndUser" says "P" canSay hasTag(path, "EndUser", "userSelection") :- isAccessPath(path).
```

For `say_hasTag("EndUser", "foo", "EndUser", "userSelection")` to be true, we will need `says_hasTag("P", "foo", "EndUser", "userSelection")` needs to be true.  Right?  We will need `ownsAccessPath` (or the `"ownership"` tag) to generate the fact for the `claim "foo" is "userSelection"` in particle "P" as follows:

```
says_hasTag("P", "foo", owner, "userSelection") :- ownsAccessPath("foo", owner").
```
or the equivalent
```
says_hasTag("P", "foo", owner, "userSelection") :- mayHaveTag("foo", owner, "ownership").
```

Essentially, `ownsAccessPath` gives us a way to _late-bind_ the owner component of a tag depending upon what the graph looks like. Consider the following graph where we have `ownsAccessPath("A", "A.ingest")`:

```
A.ingest -> P.foo
```

The following will be true:
```
says_hasTag("P", "foo", "A", "userSelection")
```

On the other hand, consider the following graph with `ownsAccessPath("A", "A.ingest")`  and `ownsAccessPath("B", "B.ingest")`:
```
A.ingest -> P.foo
B.ingest -> P.foo
```
The following will be true:
```
says_hasTag("P", "foo", "A", "userSelection")
says_hasTag("P", "foo", "B", "userSelection")
```

Reply:
@aferr any thoughts on this?

Reply:
I think I still don't see why it's needed, but you can probably move ahead with it anyway. (As far as I can tell, you can accomplish the same thing by just adding to your database: `says_hasTag("P", "ingest", "A", "userSelection"). says_hasTag("P", "ingest", "B", "userSelection").` whenever that information is available.)

Reply:
> (I think it might be that you are adding the handle connections A.ingest -> P.foo at runtime, and this is what actually solves the problem)

Yes, adding it will be part of the solution. 

> Without ownsAccessPath
(Assume we have the relevant User says P canSay hasTag(...) facts:)
> Initially with just A in the room:
we have directly saysHasTag(A, A.ingest, A, userSelection) (and hasTag(A.ingest, A, userSelection))

I think this is where there is a misunderstanding. The `userSelection` neeed not added bet on `A.ingest`, but further down the path,  say in particle `P2` of the path `A.ingest -> P1 -> P2 -> P3`.

```
particle A
  ingest: writes Foo
  claim ingest is ("A", "owner") // i.e., ownsAccessPath("A", "A.ingest")

particle P1
  in: reads Foo
  out: writes Foo

particle P2
  in: reads Foo
  out: writes Foo
  claim out is "userSelection"

particle P3
  in: reads Foo
  out: writes Foo
  check in is "userSelection"

Recipe: A -> P1 -> P2 -> P3
```

About the breakages:
> with ownsAccessPath:
We have ownsAccessPath(A, A.ingest) and as a result we prove saysHasTag(P, foo, A, audio) and saysHasTag(P, foo, A, video) even though it should only have the audio data

Note that `saysHasTag(P, foo, A, audio)` and `saysHasTag(P, foo, A, video)` will be true. However, `saysHasTag(A, foo, A, video)` will not be true `A` has allowed `P` to make that statement.  `saysHasTag(A, foo, A, video)` is needed to prove `hasTag(foo, A, video)`.  

> Where (2) breaks:
suppose A is a piece of software that has audio data for A and C and needs to input both of these to the graph involving P:

I don't completely understand this example as the edges in the data flow graph should  indicate that data from `C` goes to particle `P`. If so, there should be a place where we `ownsAccessPath` fact that will be propagated along the dataflow edges.

At this point, I don't see a way to generate datalog statements for a `claim` or a `check` of a particle without using a notion like `ownsAccessPath`. Any other solution seems to requires full understanding of the dataflow graph.  Let me move forward with this PR. We can always rollback it if I am wrong.

Reply:
OK sounds good.